### DISCUSSION TITLE
Customizing a Hab, FXGL, LibGDX client

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#304] [Ideas]
Currently, each building has a standard look and configuration, as defined in `buildings.xml` 

By definition, a habitation module is supposed to be a highly customizable building.

So where do we start to add customization to buildings such as `Lander Hab` ? 

![hab modules small](https://user-images.githubusercontent.com/1168584/103326709-79e14e00-4a06-11eb-8421-1227fcbb37bb.jpg)

**Comment by Urwumpe** [comment]
Yes. Just look at how the ISS looks inside after some years of living there - that is what habitation means, even from a psychological point of view. 

And if you want to include a more first-person perspective like commander mode, adding personal decorations and equipment to your own quarters inside a hab would also make sense to increase immersion.

As starting point, what about having a type of non-functional gear, that can be attached to walls. And once this works, we can go on giving the gear functions or psychological effects.

**Comment by mokun** [comment]
@Urwumpe 

> As starting point, what about having a type of non-functional gear, that can be attached to walls. 

So, say, we want to visualize an EVA suit. It's easy since we already have `EVASuit` class, which inherited from `Equipment` class . Now it's a matter of showing it in the Settlement Map.

But let say I need a floodlight that can illuminate the inside and outside of a building. 

And we can also install these floodlight in an outside site station or have a line of them will illuminate a road between buildings.

So, for small item such as a floodlight, an EVA suit or even a chair, is there a better way to track it instead of using the `Equipment` class to create an object out of it ? 

Btw, it's really long overdue for adding in the logics for these movable furniture inside each building.

**Comment by Urwumpe** [comment]
Well, my 2.5 cents (Euro) there: Lets assume, we have no EVA suit lying around on the ground in the base, if we do things properly. 

We have a EVA suit station, where an EVA suit can be stored, when not in use, which would be a piece of furniture in the base, with a state. This furniture object would refer to a FurnitureDefinition object for defining its properties in game as data. I think, you agree, that avoiding too many special subclasses would be better for getting things done, if we could also do things by composition there.

For making Equipment show up on top or inside furniture, we should associate it with a "VisualizationSpec" object, that defines the important data that is needed for a client to display it. For which graphic files should the client look by just naming the identifier and revision to look for, maybe with hierarchy of increasingly abstract graphic identifiers, so that clients, that not yet support the precise visual for this Equipment can still display a more generic substitute until it gets its update. 

In a perfect world, the graphics client would ask its own graphics repository server, if the graphics exist and download it, while displaying the locally existing data as temporary deputy.

For placing an Equipment on or into furniture, we need to tell the simulation, how large the item is. This could be done in the Equipment class or ideally in a "geometry outline" class, that allows it to be more compex. Next, the simulation needs to know, how this object can be placed. There could be multiple ways (laying on top of a table, hanging on a wall, using a Siemens air hook), so it makes sense to define this in an object and associate it with one or more Equipments. I have no idea yet, how to fully describe what is possible, so I don't attempt it. We should extend the enum of valid identifiers for how to place Equipment as needed. Once we know how large it is, we can think about how to store it best. Do we use a grid? Simply a volume that gets magically reduced by just the volume of the object? This could depend on how the object will be stored: Socks carefully placed flat on a table require a different layout compared to just stuffing them into a drawer - while storing a ladle in a drawer would be a pleasure for the goddess Anoia. 

Finally, I think we should wrap the Equipment actually into a "PlacedEquipment" class object when put into or onto furniture that acts as decorator pattern. This would have the elegant side effect, that we could easily define different states of the equipment without changing the equipment object. A carefully folded EVA suit would be a new PlacedEquipment object, instead of having to handle special states for equipment that can be folded. 

Maybe an EVA suit needs to be disassembled when doffed and reassembled when donned - then the Equipment EVA Suit would be created when donning it, while the Furniture actually only knows the PlacedEquipment objects that describe the components of an EVA suit and the EVA suit object would be destroyed, when doffing it and storing its parts in the EVA suit locker. In that case, the StoredEquipment could be completely unrelated to Equipment that can be used, in contrast to for example, a microscope in the lab or a microwave in the kitchen.

To be efficient, such layout decisions should only be done ONCE and kept stored as long as they are not changed. 

And finally, the client used to view the simulation should decide how to render things. An Roguelike terminal client for Mars Sim (like the console?) would not care about the graphics information at all maybe, and only handle PlacedEquipment information, or could use the VisualizationSpec just for displaying an image or a rotating 3D model of the object in a side window. While a full 3D client for Mars Sim could display the whole scene in glorious 3D using all the information the core simulation passes to it and maybe even adds some temporary data to accelerate the rendering. Or a simpler client could just ignore it completely. 

Also, the simulation engine could be made customizable to only support layout algorithms of a certain quality. Like, for increased performance, just using user predefined slots on furniture surfaces, instead of deciding by a more complex calculation, if and how a piece of equipment could be placed there - and leave it to the player setting the simulation up, how complex he likes it. If a layout calculation is not supported by the engine, its specification would be ignored in the Layout information for an object and only the supported modes will be used. And also the other way around: If we forgot to update the layout information for an old piece of equipment to support a new layout engine, only the supported modes could be used to still be able to place the object, even if more complex calculations would be supported by the engine. And just log a warning to the developers ONCE that the object needs more work.

For placing lights and other objects outside the base - we could define them as light-weight buildings, maybe just having only the most necessary information for such objects in a common super class, and define the old complex large buildings and underground structures as "new" Buildings. While storing them in a trailer or on a rover would again be the same use case as inside buildings and into or onto furniture. 

So, the actual handling of such layout decisions should be in a component object for both buildings and rovers, instead of writing the same code twice. Maybe this means that the inside of rovers can be customizable for free. Well, why not?

**Reply by mokun** [reply]
> And finally, the client used to view the simulation should decide how to render things.

But we're still using Java2D and Swing to draw things.

Guess it's about time we switch to [FXGL](https://github.com/AlmasB/FXGL) so that we can prototype graphical objects quickly, while retaining the core simulation engine.

@Urwumpe @bevans2000 

Shall we begin adopting FXGL starting v3.2.0, once we wrap up the last bit of work in fixing the airlock bug in v3.1.2 ?

**Reply by mokun** [reply]
> rotating 3D model of the object in a side window. While a full 3D client for Mars Sim could display the whole scene in glorious 3D

If we see mars-sim 3D is your future, then we may be we should consider adopting LibGDX and prettying it up with a Physics Based Rendering (PBR) shader such as [GDX-GLTF]( https://github.com/mgsx-dev/gdx-gltf). 

See this [blog](https://libgdx.com/news/2021/02/gdx-gltf) for intro.

See also this [youtube demo](https://www.youtube.com/watch?v=oma7eqXOy2M)

**Reply by Urwumpe** [reply]
> @Urwumpe @bevans2000
> 
> Shall we begin adopting FXGL starting v3.2.0, once we wrap up the last bit of work in fixing the airlock bug in v3.1.2 ?

IMHO it isn't urgent, but I wouldn't stop you from making a libGDX client, maybe I would join there later. If this client would be its own separate project (and not part of the Maven build) as created by the gdxsetup tool, you could in theory create a client that could also be installed on a smartphone and connect to a simulation via net. 

Then including a Marssim-webservice version might be a more rewarding subproject in the mars-sim project context. And the GDX client would be better in its own separate repository and development cycle.

But my own preference of development would rather be in the backend, like including remote depots and outposts. Or having a process in which the AI engineers upgrades of the existing bases, like new construction missions. And we still have no use for the ERVs (why should we, the settlers die faster than they could return to Earth).

**Reply by mokun** [reply]
@bevans2000 @Urwumpe @scud1 ,

> making a libGDX client, maybe I would join there later. If this client would be its own separate project (and not part of the Maven build) as created by the gdxsetup tool, you could in theory create a client that could also be installed on a smartphone and connect to a simulation via net.
 
Yes. LibGDX for sure can run on android phone or iOS and become a client.
 
Even FXGL can be compiled to run on smartphone as well.

> Then including a Marssim-webservice version might be a more rewarding subproject in the mars-sim project context. And the GDX client would be better in its own separate repository and development cycle.

If we adopt LibGDX or FXGL, chances are we need to switch to Android Studio IDE directly, instead of using Eclipse with its maven plug-in.

Yes, it's probably easier if we set up a separate repository.

See more at https://libgdx.com/dev/project-generation/

**Reply by bevans2000** [reply]
I agree. We have to focus on the engine. For me the goal is to make Mars sim client server. This would then allow us to have multiple clients implementations. The current Swing implementation is fit for purpose I think. However look how tightly coupled it is to the engine. Developing a 3D client would be good but if it results in another tightly coupled client then it should be delayed.
I'm working on reworking the Task logic to reduce the duplicate code and the coupling. It's going well but a bit  bigger than I intended for the first phase. Been busy with work this week so not had much time to advance it.

**Reply by mokun** [reply]
@Urwumpe @bevans2000 

See my [commit](https://github.com/mars-sim/mars-sim/commit/248df8248fde5b63c43de7bf100dcdbe17ab0614) in which I just added a new maven submodule called `mars-sim-libgdx`. 

I'm not sure if it will be automatically added to your IDE or you may have to manually import it.

I've already created a pom.xml with some basic libGDX artifacts.

But somehow it doesn't let me import `LwjglApplicationConfiguration` and `LwjglApplication` class. 

See [1](https://github.com/libgdx/libgdx/blob/master/backends/gdx-backend-lwjgl/src/com/badlogic/gdx/backends/lwjgl/LwjglApplicationConfiguration.java) and [2](https://github.com/libgdx/libgdx/blob/master/backends/gdx-backend-lwjgl/src/com/badlogic/gdx/backends/lwjgl/LwjglApplication.java).

They are needed for the desktop application but are not being recognized, despite the fact that I have followed this [post](https://stackoverflow.com/questions/63289080/trying-to-run-libgdx-with-maven-and-eclipse) to download the gdx-natives.jar package from libGDX site and manually copy the jar and the gdx64.dll explicitly onto the root level of the submodule folder. 

I've tried playing with the setup in the Build Path for mars-sim-libgdx folder but nothing works.

I don't know if libGDX has to have gradle or not.

Do let me know if you can fix the issue.

**Reply by mokun** [reply]
You can see I do have the `gdx64.dll` and the two jars in the Build Path. 

I suppose the `gdx-natives.jar` is NOT needed, since I already have the `gdx-backend-lwjgl-1.10.0.jar`

![image](https://user-images.githubusercontent.com/1168584/119927283-0187a600-bf2e-11eb-9407-d9f5975b5e0c.png)

**Reply by bevans2000** [reply]
@mokun Did you mean to commit the JAR & DLL files to git?

**Reply by mokun** [reply]
Feel free to delete that. I was trying to see if libgdx is purely mavenzied and can be loaded and run in that way

**Reply by mokun** [reply]
My question is, why aren't those two class files get imported properly ?

You can take a look at this screenshot. Both class are inside that jar file. 

![image](https://user-images.githubusercontent.com/1168584/120055956-204d7180-bfee-11eb-996f-a7cde97a9976.png)

**Reply by mokun** [reply]
Just saw you removed the scope xml tag [here](https://github.com/mars-sim/mars-sim/commit/2a0989911bf910a91e180dd38b3367891a750ae7) and it works now.

**Comment by mokun** [comment]
@Urwumpe 

> But my own preference of development would rather be in the backend, like including remote depots and outposts. 

Sure. but we can already build separate daughter mining outpost, etc. 

I guess what you want is to be remote depots to function UNIQUELY and they will supplement what's missing from its main settlement in terms of having some missions and tasks for settlers to go there to do things.

**Reply by Urwumpe** [reply]
AFAIR, this function is still blocked - we can't found new outposts during simulation that operate similar to the old ones (buildings, inventory, life support) 

Yes, the main difference for me would be setting a different strong focus on the function of the site. No permanent habitation, little recreation, lots of storage for tools and rovers, only limited automated production, no own management. They would be just "colonies" for satifying the demands of the, well, other colonies.

**Reply by mokun** [reply]
> this function is still blocked - we can't found new outposts during simulation

Fixing this now at [here](https://github.com/mars-sim/mars-sim/issues/354).

**Comment by mokun** [comment]
@Urwumpe 

> Or having a process in which the AI engineers upgrades of the existing bases, like new construction missions. 

But what are the mission aspects of expanding a typical settlement ? 

I almost feel like we need to CLEARLY DEFINE goals for players to follow through, instead of just giving them a sandbox to play whatever they want.

**Reply by Urwumpe** [reply]
I don't opposite this tutorial idea, but I think we can also have it in sandbox. 

My idea there would be that there is a feedback loop including the minds of the settlers, that specifies which demands on the settlement they have. My earlier suggestion was using a ERG model there for modelling how demands change, while maybe learning from Maslow, what kinds of specific demands could exist. Additionally, for the beginning, we could also have Earth specifying their demands and assisting in survival. Which might be a good extension point for scripted tutorials as well.

**Reply by mokun** [reply]
> My earlier suggestion was using a ERG model there for modelling how demands change, while maybe learning from Maslow, what kinds of specific demands could exis

So, you see the value of giving each settler on Mars *desires*, much like the **Sims**, right ?

Say, under the *Command Mode*, one of the goals of the player is to help fulfill these individual *desires* under the Maslow Hierarchy.

When a desire is met, the player can get more leadership points, for instance. 

The settlers who have the desires fulfilled will also be more *motivated* to do the player's bidding, etc.

But you're saying there's also a layer of the overall settlement's demand that the player needs to meet. 

If fulfilled successfully within a certain period of time, the player can get promoted or score more or something.

**Comment by mokun** [comment]
@Urwumpe 

> And we still have no use for the ERVs (why should we, the settlers die faster than they could return to Earth).

Yes since MarsOne is now out of the picture, we need to piece together in how the ERV can be used, instead of making it like a resource processing plant.

But to do so, we need to take a look at the current weight of each item and sum them up to see if it makes sense to bring them to Mars and if it's feasible on the kind of payload space we envision...

May be only after we add in new codes to check for the space and weight requirement, then we can work on using ERV to return people and items ?! 

What do you think ?

**Reply by mokun** [reply]
@bevans2000 ,

What change in the current code architecture do you envision to enable the ERV to return back to Earth ?

**Reply by bevans2000** [reply]
The key aspects when do the people go out of scope. Once the ERV takes off then they have to be removed from the updating pulse logic.

**Reply by mokun** [reply]
> The key aspects when do the people go out of scope. Once the ERV takes off then they have to be removed from the updating pulse logic.

Sure.

What if we create a virtual location called earth as well as the moon base ? 

Then we can build some rules for the settlement on Mars to trade resources with these two super remote locations 

Also, we can think of offloading people there for vacation and/or business or work visit , or even setting up contractual agreement to (re)hire them to come back and forth to Mars to work. 

This will be more realistic for corporations that may be expanding off-world locations

**Reply by Urwumpe** [reply]
> 
> What do you think ?

Well, maybe it would already be enough to start at the ERV with two functions:

1. Source for produced fuel for return trip. Has to be fueled for return.
2. Return settlers to Earth after their time on Mars.

For the first one, we should have pretty much everything in place already.
For the second function, the Settlers would first need a specification of how and how long they stay on Mars. Maybe give them different "Mars tickets", with the types of possible Mars tickets initially being configured in XML.

Next, we should give them tasks at different intensity and priority to prepare for leaving Mars as the end of their stay approaches.  Packing their bags. Preparing the ERV, getting suited up, leave mars. Missions and new research projects would have to be banned for them as soon as the departure approaches. 

At one point, there might be more or less seats available in a ERV for the astronauts to leave Mars, maybe we need some kind of smarter logistics there then, aside of just letting a settler spend more time on Mars.

**Reply by mokun** [reply]
> For the second function, the Settlers would first need a specification of how and how long they stay on Mars. Maybe give them different "Mars tickets", with the types of possible Mars tickets initially being configured in XML.

Sure. To do so, we need to figure out the REASON for anyone coming to Mars, or going to the Moon. 

Their rationale will drive their actions, right ?
